File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/RequireTemplateGroupsRule.java
package ai.harness.maven;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.List;

/**
 * Rule that requires specific template groups to exist.
 */
public class RequireTemplateGroupsRule implements HarnessCheckRule {
    private static final String CATEGORY = "requireTemplateGroups";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public List<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        JsonNode catNode = manifest.get(CATEGORY);
        String targetRoot = catNode.get("parameters").get("targetRoot").asText();
        List<String> groups = HarnessCheckHelper.extractPaths(catNode.get("parameters").get("groups"));
        String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return groups.stream()
                .filter(g -> !HarnessCheckHelper.isSafeDirectory(root, root.resolve(targetRoot).resolve(g)))
                .map(g -> new Finding(severity, CATEGORY, "missing template group: " + targetRoot + "/" + g))
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/RequireTemplateGroupsRule.kt
package ai.harness.gradle

import kotlinx.serialization.json.JsonObject
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isDirectory
import kotlin.io.path.isSymbolicLink

/**
 * Rule that requires template groups to exist as directories.
 */
class RequireTemplateGroupsRule : HarnessCheckRule {
	override fun applies(manifest: JsonObject): Boolean {
		val category = "requireTemplateGroups"
		val catObj = manifest[category]?.jsonObject ?: return false
		val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
		return enabled
	}

	override fun validate(manifest: JsonObject, root: Path, psiResults: HarnessPsiResults?): List<Finding> {
		val category = "requireTemplateGroups"
		val severity = HarnessCheck.Companion.severityOf(manifest, category)
		val catObj = manifest[category]?.jsonObject ?: return emptyList()
		val parametersObj = catObj["parameters"]?.jsonObject ?: return emptyList()
		val targetRoot = HarnessCheck.Companion.stringFrom(parametersObj, "targetRoot")
		val groups = HarnessCheck.Companion.stringArrayFrom(parametersObj, "groups")
		return buildSet<Finding> {
			groups.forEach { group ->
				val p = root / "$targetRoot/$group"
				when {
					p.isSymbolicLink() -> add(Finding(Severity.ERROR, category, "symlink directory is not allowed: $targetRoot/$group"))
					!p.isDirectory() -> add(Finding(severity, category, "missing template group: $targetRoot/$group"))
				}
			}
		}.toList()
	}
}
